### ISSUE TITLE
Allow `SetRequired` to specify internal properties (aka optic)

### ISSUE THREAD

**Issue by Lonli-Lokli** [sindresorhus/type-fest#569]
Assuming we have 
```ts
type FormValue = {
  person: {
   city?: string;
  }
}
```

SetRequired do not allow creation of FormValuee with city set as required. I belive it can be done with template literal types,

eg
```ts
type DeepKeyOf<T> = (
  [T] extends [never] ? "" :
  T extends object ? (
    { [K in Exclude<keyof T, symbol>]:
      `${K}${undefined extends T[K] ? "?" : ""}${DotPrefix<DeepKeyOf<T[K]>>}` }[
    Exclude<keyof T, symbol>]
  ) : ""
) extends infer D ? Extract<D, string> : never;

type DotPrefix<T extends string> = T extends "" ? "" : `.${T}`;
```

**Comment by tommy-mitchell** [comment]
So essentially a `SetRequiredDeep` type?

**Comment by Lonli-Lokli** [comment]
I don't think it makes  sense to add new one, I think it's better to upgrade existing, as all other code will continue to be valid and type-safe.
The only difference is that it will be able to support nested props

**Comment by Lonli-Lokli** [comment]
Btw I think https://github.com/sindresorhus/type-fest/blob/main/source/set-optional.d.ts can be updated too

**Comment by Lonli-Lokli** [comment]
@tommy-mitchell I can send PR if you think it's ok

**Comment by tommy-mitchell** [comment]
That’s up to Sindre, I was just making a suggestion based on the other types in this library.

**Comment by Lonli-Lokli** [comment]
@sindresorhus what do you think?

**Comment by sindresorhus** [comment]
So you want `SetRequired<Foo, 'person.city'>` to work, correct? That would be a breaking change as it's ambiguous whether `person.city` is a key path or just a key name. `person.city` is a perfectly valid key name.

I think it would be better to make a `SetRequiredDeep` that by default makes all keys recursively required and if you specify a key path, it makes a specific deep key required. @tommy-mitchell What do you think?

**Comment by tommy-mitchell** [comment]
> I think it would be better to make a `SetRequiredDeep` that by default makes all keys recursively required and if you specify a key path, it makes a specific deep key required.

I think that fits with the rest of `type-fest` and has an intuitive usage.

**Comment by Lonli-Lokli** [comment]
So here is draft playground for keys https://tsplay.dev/Wzky4w

**Comment by sindresorhus** [comment]
@Lonli-Lokli Thanks for elaborating. My only concern left for having it in `SetRequried` is that it may affect type-checking performance. We had this problem in `Simplify` and had to split the deep handling (it even affected non-deep usage) into a separate type.

**Comment by sindresorhus** [comment]
> So here is draft playground for keys https://tsplay.dev/Wzky4w

The key path handling is similar to https://github.com/sindresorhus/type-fest/pull/409/files and `Get`. Would be nice to somehow unify all that.

**Comment by tommy-mitchell** [comment]
> Would be nice to somehow unify all that.

I've been working on a `Paths` type

**Comment by Lonli-Lokli** [comment]
I will just leavee it here as a some kind of sample

```ts
type StringOrNumKeys<TObj> = keyof TObj & (string | number);

type NestedPath<TValue, Prefix extends string> =
    TValue extends object
    ? `${Prefix}.${NestedFieldPaths<TValue>}`
    : never;

export type NestedFieldPaths<TData> = {
    [TKey in StringOrNumKeys<TData>]: `${TKey}` | NestedPath<TData[TKey], `${TKey}`>;
}[StringOrNumKeys<TData>];
```

**Comment by Lonli-Lokli** [comment]
Let me up this issue, is there any plans to implement it in the nearest future?

**Comment by som-sm** [comment]
Closing this since we now have a [`SetRequiredDeep` type](https://github.com/sindresorhus/type-fest/pull/939).
